keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,364 @@
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: some-pod
+spec:
+  containers:
+    ...
+  volumes:
+      - name: vol
+        csi:
+          driver: inline.storage.kubernetes.io
+          volumeAttributes:
+              foo: bar
+```
+
+## Proposal
+This proposal introduces the support of both persistent and ephemeral-style volumes nested within a pod spec definition. The following highlights the major themes of this proposal:
+
+* Support for both persistent and ephemeral volumes
+* Propose a way to clearly distinguish between persistent and ephemeral volumes.
+* Persistent inline CSI volumes must be pre-provisioned and manually de-provisioned.
+* Persistent inline CSI volumes will participate in storage flows including attach, stage, mount, unmount, unstage, and detach.
+* Persistent inline volumes must use the handle of pre-provisioned volumes
+* An ephemeral volume is created and deleted automatically following its pod's lifecycle
+* Inline ephemeral volumes receive a generated volume handle
+
+### Inline volumes
+Normally, volume operations can be triggered from PV/PVC or when from a pod spec.  Currently, CSI only supports volumes that originate from PV/PVC. When a volume is embedded inside a pod spec, it will be referred to as `inline` in this document.  Volumes that originates from inline pod specs can support two modes:
+
+* Persistent mode - the volume follows a persistent lifecycle independent of the pod where it is used.
+* Ephemeral mode - the volume follows the lifecycle of the pod where it is used.
+
+A driver may be able to support either PV/PVC-originated or pod spec originated volumes (inline).  The following is important to keep in mind:
+* When a volume originates from PV/PVC, the driver will receive all volume related CSI calls (normal operation).
+* When a volume originates from a pod spec (inline), the CSI driver will only receive a limited number of calls (see below).
+
+### Inline CSI lifecycle setting
+A CSI driver will have to indicate the inline volume lifecycle, that it supports, as either `persistent` or `ephemeral`.  This setting will be stored in a [`CSIDriver` configuration CRD](https://github.com/kubernetes/enhancements/issues/594) and will be checked at runtime by Kubelet and external CSI components for enforcement.  The setting is to inform the Kubelet and other CSI components how to process inline volume information for proper setup.
+
+#### Persistent inline mode
+For CSI drivers that indicate they support persistent inline volumes, the `CSIDriver.lifecycle` value should be set to `persistent` (default of if not provided).  When a driver is used to handle volumes from an inline context: 
+* Volumes are assumed to be originated from an inline pod spec.
+* The `volumeHandle` value must be generated ahead of time and manually entered in the pod spec.
```

Comment:
runtime's pretty easy to implement. api validation would be harder? can an api validator lookup cr's to validate against? can we add to the existing api validation logic or would that need to be an external webhook thing?

Reply:
@liggitt this will have to happen at runtime since volumeHandle can support two modes one of which does not require it when ephemeral.  The api validators don't usually call the API server (or at least i haven't seen one), but if possible or doable, then the validator will check the CRD for proper application.